interceptor.go

```diff
@@ -0,0 +1,90 @@
+// +build !js
+
+package webrtc
+
+import (
+	"io"
+
+	"github.com/pion/rtcp"
+	"github.com/pion/rtp"
+)
+
+// WriteRTP is used by Interceptor.BindLocalTrack.
+type WriteRTP func(p *rtp.Packet, attributes map[interface{}]interface{}) (int, error)
+
+// ReadRTP is used by Interceptor.BindRemoteTrack.
+type ReadRTP func() (*rtp.Packet, map[interface{}]interface{}, error)
+
+// WriteRTCP is used by Interceptor.BindWriteRTCP.
+type WriteRTCP func(pkts []rtcp.Packet, attributes map[interface{}]interface{}) (int, error)
+
+// ReadRTCP is used by Interceptor.BindReadRTCP.
+type ReadRTCP func() ([]rtcp.Packet, map[interface{}]interface{}, error)
+
+// Interceptor can be used to add functionality to you PeerConnections by modifying any incoming/outgoing rtp/rtcp
+// packets, or sending your own packets as needed.
+type Interceptor interface {
+
+	// BindReadRTCP lets you modify any incoming RTCP packets. It is called once per sender/receiver, however this might
+	// change in the future. The returned method will be called once per packet batch.
+	BindReadRTCP(read ReadRTCP) ReadRTCP
+
+	// BindWriteRTCP lets you modify any outgoing RTCP packets. It is called once per PeerConnection. The returned method
+	// will be called once per packet batch.
+	BindWriteRTCP(write WriteRTCP) WriteRTCP
+
+	// BindLocalTrack lets you modify any outgoing RTP packets. It is called once for per LocalTrack. The returned method
+	// will be called once per rtp packet.
+	BindLocalTrack(ctx *TrackLocalContext, write WriteRTP) WriteRTP
+
+	// UnbindLocalTrack is called when the Track is removed. It can be used to clean up any data related to that track.
+	UnbindLocalTrack(ctx *TrackLocalContext)
+
+	// BindRemoteTrack lets you modify any incoming RTP packets. It is called once for per RemoteTrack. The returned method
+	// will be called once per rtp packet.
+	BindRemoteTrack(ctx *TrackRemoteContext, read ReadRTP) ReadRTP
```

Comment:
Don't we want to split `Interceptor` into `InterceptorLocal` and `InterceptorRemote`? I guess most of the interceptor implementation implements only one of local/remote.

Reply:
There is a possibility that someone may want both! I was thinking it would be nice if `interceptor.Stats` covered both paths. I am sure we will have more as we build them out.

The nice thing is that a user doesn't need to define everything, they can embed a `interceptor.NoOp` and then only define the things they care about.

Reply:
I was thinking about this a lot. Even thought most interceptors either only handle one of outgoing rtp/incoming rtp, or can be split easily into two (eg: NackRemote and NackLocal, TransportCCRemote and TransportCCLocal), there are some that need to process both incoming and outhoing rtp streams.

For example a SenderReport can include ReceptionReports. In order to generate this, it will need to monitor both incoming and outgoing rtp streams. Even with Local/Remote separation, this could be solved by passing the same struct that implements both interfaces as both a Remote and Local interceptor, but that would mean the interceptor is registered twice, and all common methods (RTCP handling, Close) would be called twice.

So in short:
1 interface:
- con: lot of interceptors will not implemented certain methods (but this can be easily achived by embedding the NoOpInterceptor)
- pro: certain interceptors that do need all the methods are easier to implement
Separate Local/Remote:
- con: certain interceptor would be very difficult to implement
- pro: slightly smaller interfaces

Based on this I decided to go with a single interface.

Even with a single interface, it's possible to do the separation (eg: NackRemote vs NackLocal), so people can use only one of them if they want to.

Reply:
I am in support of the single interface! I think the pros outweigh the conns in this case!